### ISSUE TITLE
kubernetes deployment yaml for docker for desktop (mac and windows)

### ISSUE THREAD

**Issue by scriptjs** [rancher/rancher#14209]
I realize there is work ongoing for a helm chart that can be deployed on mac or windows docker for desktop. In one of the previous meetups, a question was asked about this. Darren did pointed to the work with the helm chart that is ongoin but it currently has an issue related to fqdn.

I presently have a 4GB ubuntu vm working against the single node mac kubernetes cluster which works. Combined, this remains heavy on my local mac with 16GB ram. I'd like to be running the single node cluster in 4 GB RAM as opposed to 8GB.

While the issues with the chart are being resolved, is there a kubernetes yaml that can be deployed in the interim from a gist or other source? Many thanks.

**Comment by jgreat** [comment]
If you're up for a test you can try out this pre-release chart on the 2.0.4 branch: https://github.com/jgreat/server-chart/

An overview of the process is here, but you will want to use the repo above to install the rancher chart.
https://rancher.com/blog/2018/2018-05-18-how-to-run-rancher-2-0-on-your-desktop/

TL;DR Instructions:
1. Enable kubernetes in docker-for-desktop
1. helm init...
1. Install the nginx ingress controller:  `helm install stable/nginx-ingress --name ingress-nginx --namespace ingress-nginx`
1. Clone the repo and check out the 2.0.4 branch
1. `cd server-charts/rancher`
1. Install rancher using the local chart `helm install ./ ...`

See the README.md for rancher configuration details, but the only thing you probably need to set is `hostname`

**Comment by scriptjs** [comment]
@jgreat Many thanks for your reply. I have identified the steps I took below. After installation, I navigated to the https address with the following.  
```
default backend - 404
```
Inspecting the cert shows that the cert is for ingress.local for Kubernetes. I have set up the kubernetes dashboard so am able to provide additional information as needed to help debug the issue. Many thanks.

=====

Verify context
```
kubectl get nodes
NAME                 STATUS    ROLES     AGE       VERSION
docker-for-desktop   Ready     master    50m       v1.10.3
```
Install tiller service account
```
$ kubectl -n kube-system create serviceaccount tiller
serviceaccount "tiller" created
```
Install tiller ClusterRoleBinding
```
$ kubectl create clusterrolebinding tiller --clusterrole cluster-admin --serviceaccount=kube-system:tiller
clusterrolebinding.rbac.authorization.k8s.io "tiller" created
```
Initialize tiller service
**Note:** I have replaced the user of the home directory with <user> from the actual output below:
```
$ helm init --service-account tiller
$HELM_HOME has been configured at /Users/<user>/.helm.

Tiller (the Helm server-side component) has been installed into your Kubernetes Cluster.

Please note: by default, Tiller is deployed with an insecure 'allow unauthenticated users' policy.
For more information on securing your installation see: https://docs.helm.sh/using_helm/#securing-your-helm-installation
Happy Helming!
```

Install nginx Ingress
```
$ helm install stable/nginx-ingress --name ingress-nginx --namespace ingress-nginx
NAME:   ingress-nginx
LAST DEPLOYED: Thu Jun 28 15:14:06 2018
NAMESPACE: ingress-nginx
STATUS: DEPLOYED

RESOURCES:
==> v1/Service
NAME                                         TYPE          CLUSTER-IP      EXTERNAL-IP  PORT(S)                     AGE
ingress-nginx-nginx-ingress-controller       LoadBalancer  10.109.106.189  localhost    80:32513/TCP,443:32533/TCP  0s
ingress-nginx-nginx-ingress-default-backend  ClusterIP     10.99.167.100   <none>       80/TCP                      0s

==> v1beta1/PodDisruptionBudget
NAME                                         MIN AVAILABLE  MAX UNAVAILABLE  ALLOWED DISRUPTIONS  AGE
ingress-nginx-nginx-ingress-controller       1              N/A              0                    0s
ingress-nginx-nginx-ingress-default-backend  1              N/A              0                    0s

==> v1/Pod(related)
NAME                                                          READY  STATUS             RESTARTS  AGE
ingress-nginx-nginx-ingress-controller-fc784c8c9-htzdh        0/1    ContainerCreating  0         0s
ingress-nginx-nginx-ingress-default-backend-7fbcd89c6c-ncmcp  0/1    ContainerCreating  0         0s

==> v1/ConfigMap
NAME                                    DATA  AGE
ingress-nginx-nginx-ingress-controller  1     0s

==> v1/ServiceAccount
NAME                         SECRETS  AGE
ingress-nginx-nginx-ingress  1        0s

==> v1beta1/ClusterRole
NAME                         AGE
ingress-nginx-nginx-ingress  0s

==> v1beta1/Role
ingress-nginx-nginx-ingress  0s

==> v1beta1/ClusterRoleBinding
NAME                         AGE
ingress-nginx-nginx-ingress  0s

==> v1beta1/RoleBinding
NAME                         AGE
ingress-nginx-nginx-ingress  0s

==> v1beta1/Deployment
NAME                                         DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
ingress-nginx-nginx-ingress-controller       1        1        1           0          0s
ingress-nginx-nginx-ingress-default-backend  1        1        1           0          0s


NOTES:
The nginx-ingress controller has been installed.
It may take a few minutes for the LoadBalancer IP to be available.
You can watch the status by running 'kubectl --namespace ingress-nginx get services -o wide -w ingress-nginx-nginx-ingress-controller'

An example Ingress that makes use of the controller:

  apiVersion: extensions/v1beta1
  kind: Ingress
  metadata:
    annotations:
      kubernetes.io/ingress.class: nginx
    name: example
    namespace: foo
  spec:
    rules:
      - host: www.example.com
        http:
          paths:
            - backend:
                serviceName: exampleService
                servicePort: 80
              path: /
    # This section is only required if TLS is to be enabled for the Ingress
    tls:
        - hosts:
            - www.example.com
          secretName: example-tls

If TLS is enabled for the Ingress, a Secret containing the certificate and key must also be provided:

  apiVersion: v1
  kind: Secret
  metadata:
    name: example-tls
    namespace: foo
  data:
    tls.crt: <base64 encoded cert>
    tls.key: <base64 encoded key>
  type: kubernetes.io/tls
```
Install cert manager
```
$ helm install stable/cert-manager --name cert-manager --namespace kube-system
NAME:   cert-manager
LAST DEPLOYED: Thu Jun 28 15:20:00 2018
NAMESPACE: kube-system
STATUS: DEPLOYED

RESOURCES:
==> v1/ServiceAccount
NAME                       SECRETS  AGE
cert-manager-cert-manager  1        8s

==> v1beta1/CustomResourceDefinition
NAME                               AGE
certificates.certmanager.k8s.io    4s
clusterissuers.certmanager.k8s.io  0s
issuers.certmanager.k8s.io         0s

==> v1beta1/ClusterRole
cert-manager-cert-manager  0s

==> v1beta1/ClusterRoleBinding
NAME                       AGE
cert-manager-cert-manager  0s

==> v1beta1/Deployment
NAME                       DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
cert-manager-cert-manager  1        1        1           0          0s

==> v1/Pod(related)
NAME                                        READY  STATUS             RESTARTS  AGE
cert-manager-cert-manager-866d8dd8c9-6m8lp  0/1    ContainerCreating  0         0s


NOTES:
cert-manager has been deployed successfully!

In order to begin issuing certificates, you will need to set up a ClusterIssuer
or Issuer resource (for example, by creating a 'letsencrypt-staging' issuer).

More information on the different types of issuers and how to configure them
can be found in our documentation:

https://cert-manager.readthedocs.io/en/latest/reference/issuers.html

For information on how to configure cert-manager to automatically provision
Certificates for Ingress resources, take a look at the `ingress-shim`
documentation:

https://cert-manager.readthedocs.io/en/latest/reference/ingress-shim.html
```
Set up DNS:
**Note:** I replaced my actual domain name with <my.domain> below:
Logged in to DNS provider and set an A record for rancher-dev.<my.domain>

Install rancher 
**Note:** I have replaced my actual email address with <me@my.domain> and my domain name with <my.domain> in the command and actual output below:
```
$ helm install ./ --name rancher --namespace cattle-system --set hostname=rancher-dev.<my.domain> --set ingress.tls.source=letsEncrypt --set letsEncrypt.environment=staging --set letsEncrypt.email=<me@my.domain>
NAME:   rancher
LAST DEPLOYED: Thu Jun 28 15:28:30 2018
NAMESPACE: cattle-system
STATUS: DEPLOYED

RESOURCES:
==> v1/ClusterRoleBinding
NAME     AGE
rancher  0s

==> v1/Service
NAME     TYPE       CLUSTER-IP    EXTERNAL-IP  PORT(S)  AGE
rancher  ClusterIP  10.107.22.19  <none>       80/TCP   0s

==> v1beta1/Deployment
NAME     DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
rancher  1        1        1           0          0s

==> v1beta1/Ingress
NAME     HOSTS                     ADDRESS  PORTS  AGE
rancher  rancher-dev.devscript.io  80, 443  0s

==> v1alpha1/Issuer
NAME     AGE
rancher  0s

==> v1/Pod(related)
NAME                      READY  STATUS             RESTARTS  AGE
rancher-5985497fc9-nwvk6  0/1    ContainerCreating  0         0s

==> v1/ServiceAccount
NAME     SECRETS  AGE
rancher  1        0s


NOTES:
Rancher Server has been installed.

Check out our docs at https://rancher.com/docs/rancher/v2.x/en/

Browse to https://rancher-dev.<my.domain>

Happy Containering!

```

**Comment by jgreat** [comment]
Sometimes cert-manager/LetsEncrypt takes a while to issue a cert. 

Check the status of the Ingress in cattle-system make sure the events show that the CreateCertificate is complete and successful

```
kubectl -n cattle-system describe ingress
...
Events:
  Type    Reason             Age              From                      Message
  ----    ------             ----             ----                      -------
  Normal  CREATE             4m               nginx-ingress-controller  Ingress cattle-system/rancher
  Normal  CREATE             4m               nginx-ingress-controller  Ingress cattle-system/rancher
  Normal  CREATE             4m               nginx-ingress-controller  Ingress cattle-system/rancher
  Normal  CreateCertificate  4m               cert-manager              Successfully created Certificate "tls-rancher-ingress"
  Normal  UPDATE             4m (x2 over 4m)  nginx-ingress-controller  Ingress cattle-system/rancher
  Normal  UPDATE             4m (x2 over 4m)  nginx-ingress-controller  Ingress cattle-system/rancher
  Normal  UPDATE             4m (x2 over 4m)  nginx-ingress-controller  Ingress cattle-system/rancher
```

If you don't have a successful create check the status of the Certificate.

```
kubectl -n cattle-system describe certificate
```

That should give you some direction in your troubleshooting.

**Comment by scriptjs** [comment]
@jgreat I have a followup question regarding the helm chart. On docker for mac, the docker subnet is 192.168.65.0

When I deploy a workload, links available under a NodePort workload fail. These look like 192.168.65.0:3xxxx. That said, I am able to see the NodePort workload at localhost:3xxxx and 127.0.0.1:3xxxx. My goal was to deploy local workloads that I could see from port 80 and 443 on my local machine using my wildcard domain ie. https://drone.my.domain

So there appear to be two issues to resolve to geting there.

1. I'll need to change ports in the helm chart to deploy rancher to alternative ports to free port 80 and 443 on my local host using the single node provided by kubernetes.
2. I need to determine whether not being able to view a workload a 192.168.65.0:3xxxx is a bug or can be resolved through additional configuration.

I have looked at the deployment and ingress where ports are defined. Could you suggest where the changes ought to be made to alter the rancher ports to use 8080 and 8443 instead. Is 2 a bug? Many thanks.

**Comment by jgreat** [comment]
A little bit of clarification on the intentions here. The blog entry was designed that you run Rancher server on the docker-for-desktop kubernetes, not an extra VM.  If you install rancher in the d4d k8s "cluster" you can also create and manage workloads on that "local" that shows up by default.

We're starting to get into how Docker for Desktop sets up kubernetes, and that is a bit of a black box. I couldn't really find any documentation beyond the basic starting k8s up.

I noticed that it does some magic for you and essentially sets up a k8s cloud-provider/controller so you can use a LoadBalancer type Service to publish the ports on your desktop localhost.  You may need to use a LoadBalancer type service instead of NodePort, because the NodePorts will just be on the d4d VM.

For advice on configuration for your specific setup, I'd suggest checking out our forums or joining the community slack channel.

**Comment by scriptjs** [comment]
@jgreat Hi. To clarify, I am speaking about a workload deployed in local. For example, setting up an nginx as a workload Nodeport will provide you with a link when deployed. Click on the link and you cannot navigate to see the nginx start page, but it is instead  accessible on localhost with the TCP port determined by rancher server.

I'll look more closely at the settings in the helm chart to see what needs to be changed. My feeling is that the http and https ports should be templated to enable custom values to be used.

Thank you for your work on the chart, it is quite helpful. I also quite like that you can reset kubernetes completely with d4d also – quite good for some experiments. In any case hoping to make it into a convenient local environment.

**Comment by rgunn** [comment]
@scriptjs I came across this as well but it was after already using Docker for Mac prior to enabling Kubernetes and installing Rancher, after which exactly the same behaviour you described applies. You cannot access container IPs directly on Docker for Mac and localhost is used with port forwarding. See https://docs.docker.com/docker-for-mac/networking/#known-limitations-use-cases-and-workarounds

**Comment by scriptjs** [comment]
I have determined what would be needed to have this work natively, but it needs some attention. https://github.com/rancher/rancher/issues/14358

In the interim, I am continuing with other local virtualization.
